docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,864 @@
+    }
+
+    interface Builder {
+        /**
+         * The custom S3AsyncClient this transfer manager will use to make calls
+         * to S3.
+         */
+        Builder s3client(S3AsyncClient s3Client);
+
+        /**
+         * The configuration for this transfer manager.
+         */
+        Builder configuration(TransferManagerConfiguration configuration);
+
+        /**
+         * Add a progress listener to the currently configured list of
+         * listeners.
+         */
+        Builder addProgressListener(TransferProgressListener progressListener);
+
+        /**
+         * Set the list of progress listeners, overwriting any currently
+         * configured list.
+         */
+        Builder progressListeners(Collection<? extends TransferProgressListener> progressListeners);
+
+        S3TransferManager build();
+    }
+}
+
+/**
+ * Configuration object for {@link S3TransferManager}.
+ */
+public interface TransferManagerConfiguration {
+    /**
+     * The max number of requests the Transfer Manager will have at any point in
+     * time. This must be less than or equal to the max concurrent setting on
+     * the S3 client.
+     */
+    public Integer maxConcurrency();
+
+    /**
+     * The aggregate max upload rate in bytes per second over all active upload
+     * transfers. The default is unlimited.
+     */
+    public Long maxUploadBytesPerSecond();
+
+    /**
+     * The aggregate max download rate in bytes per second over all active
+     * download transfers. The default value is unlimited.
+     */
+    public Long maxDownloadBytesPerSecond();
+
+    /**
+     * The configuration for multipart uploads.
+     */
+    public MultipartUploadConfiguration multipartUploadConfiguration();
+
+    /**
+     * The configuration for multipart downloads.
+     */
+    public MultipartDownloadConfiguration multipartDownloadConfiguration();
+}
+
+/**
+ * Configuration object for multipart downloads.
+ */
+public interface MultipartDownloadConfiguration {
+    /**
+     * Whether multipart downloads are enabled.
+     */
+    public Boolean enableMultipartDownloads();
+
+    /**
+     * The minimum size for an object to be downloaded in multiple parts.
+     */
+    public Long multipartDownloadThreshold();
+
+    /**
+     * The maximum number of parts objects are to be downloaded in.
+     */
+    public Integer maxDownloadPartCount();
+
+    /**
+     * The minimum size for each part.
+     */
+    public Long minDownloadPartSize();
+}
+
+/**
+ * Configuration object for multipart uploads.
+ */
+public final class MultipartUploadConfiguration {
+    /**
+     * Whether multipart uploads should be enabled.
+     */
+    public Boolean enableMultipartUploads();
+
+    /**
+     * The minimum size for an object to be uploaded in multipe parts.
+     */
+    public Long multipartUploadThreshold();
+
+    /**
+     * The minimum size for each uploaded part.
+     */
+    public Long minUploadPartSize();
+}
```

Comment:
No Integer maxUploadPartCount?

Reply:
Sounds good. Lifted the settings initially from v1, and it doesn't have that config!